### DISCUSSION TITLE
Java Spring em Clojure

### DISCUSSION THREAD

**Discussion by wandersoncferreira** [clj-br/forum#23] [Geral]
Nunca usei o Spring e fiquei curioso em saber como seria uma aplicação Spring em Clojure. Consegui escrever as classes sem muito problema:

```clojure
(ns spring.rest.hello
  (:import [org.springframework.web.bind.annotation RestController RequestMapping])
  (:gen-class
   :name ^{org.springframework.web.bind.annotation.RestController {}} spring.rest.HelloController
   :methods [[^{org.springframework.web.bind.annotation.RequestMapping ["/"]} index [] String]]))


;; Código original em Java

;; import org.springframework.web.bind.annotation.RestController;
;; import org.springframework.web.bind.annotation.RequestMapping;

;; @RestController
;; public class HelloController {

;;     @RequestMapping("/")
;;     public String index() {
;;         return "Ola! Spring Boot!";
;;     }

;; }
;;

(defn -index
  [this]
  "Olaa! Clojure!")

```

Porém, não consegui rodar o projeto porque aparentemente o Spring usa uma configuração de JAR diferente e tem seu próprio launcher. Essa é a especificação do JAR do spring https://docs.spring.io/spring-boot/docs/current/reference/html/executable-jar.html#executable-jar.nested-jars.jar-structure ... 

Esse projeto tem um exemplo completo usando leiningen https://github.com/scotthaleen/clojure-spring-cloud e plugins para converter um JAR "normal" em um JAR-spring.

É realmente impressionante a quantidade de código para fazer coisas tão simples.

**Comment by ericdallo** [comment]
Bem legal!

Eu trabalhei alguns anos com spring e acho muito bom pra java, mas não consigo ver usando em Clojure hehe, spring usa muitas annotations e a não ser que alguém criasse um wrapper em Clojure, usar manualmente me parece meio complexo apesar do seu exemplo ser simples, acredito que usando outras features do spring deve ficar bem complexo e difícil de organizar sem um rapper Clojure.

Mas parabéns pela iniciativa, quem sabe um wrapper não vale a POC, spring é bem poderoso na comunidade java

**Comment by avelino** [comment]
> ajudando pessoas que chegar até essa thread futuramente (Google pode indexar)

O @jakubriegel fez um projeto usando `spring boot` e `lein` vale ver o repositório:
https://github.com/jakubriegel/clojure-spring

> _@jakubriegel sorry to tag you in a portuguese content, I'm telling you that you made an example project using `spring` and `lein`_
